Update installer (resolves DEP-21)

jessebrennan opened the pull request:
This updates the installer with the three new options that are necessary to access the whitelist.

jessebrennan committed:
Add options to install script (resolved #48)

The whitelist checking depends on three new parameters
- whitelist name
- an email for help
- the name of the project
This code adds support for adding these as environment variables into
the container.

jessebrennan commented:
I'm not really sure why there are merge conflicts, I might have done some git foo(bar) and screwed things up. @coverbeck could you review?

coverbeck reviewed (changes requested):
1. Can you please resolve conflicts
2. You need to use these in prod.yml and dev.yml.

coverbeck commented:
Your commit message should refer to the GitHub issue, not the JIRA, at least in the first line.

It would also be nice if the commit message gave a little more detail so somebody could know what changed without having to look at the code change.

coverbeck commented on install_bootstrap:
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
 #      fi
 
-      
+      project_name='project_name'
```
This will only come into play if there is an email_whitelist_name.

I think you should ask for the email_whitelist_name first. If the user does not enter one, then you shouldn't prompt for this nor for the email.

jessebrennan replied:
Good point. I've reorganized.

coverbeck commented on install_bootstrap:
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
+      ask_question "What is a contact email in case there are problems authenticating?" "$CONTACT_EMAIL" "contact email" $contact_email
+
+      email_whitelist_name='email_whitelist_name'
+      ask_question "What is the name of the AWS Secret that contains the whitelist? (e.g. commons/dev/whitelist)" "$EMAIL_WHITELIST_NAME" "email whitelist name" $email_whitelist_name
```
We want to use this as a flag to indicate whether authentication is required to access the instance. The prompt should reflect that, e.g., something like like "leave blank if you do not want to enforce authorization".

jessebrennan replied:
This is also fixed now.

coverbeck commented on install_bootstrap:
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
+      ask_question "What is the name of the project for this instance (e.g. Stage, Commons, AnVIL, etc.)? This name will be used for error messages." "$PROJECT_NAME" "project name" $project_name
+
+      contact_email='contact_email'
+      ask_question "What is a contact email in case there are problems authenticating?" "$CONTACT_EMAIL" "contact email" $contact_email
```
I would say `authorizing` instead of `authenticating`.

jessebrennan replied:
Good catch!

coverbeck reviewed (changes requested):
There are two other things in the ticket:

1. Google client ids, but that is optional, so you can leave it out for now.
2. Generating a secret for cookie encryption/decryption. I think you should tackle that one. There is a generatePassword function in install_boostrap that you can invoke.

Also, can you please update the Collecting Information section of the README with a little more information about these.

I know you said you didn't like how the README.md presents information, so if you want to refactor it overall as part of this, that would be fine too... Or just fit it into the existing format, and create another ticket to update the README.

Thanks!

jessebrennan committed:
Make formatting and indentation consistent

jessebrennan committed:
Ask about authorization and set vars accordingly

Now the script will ask the user if they want to install with
authorization an a whitelist. If they do then there are follow-up
questions. If they don't, then the email_whitelist_name variable
is set to '' to indicate no authentication will be used.

jessebrennan committed:
Update gitignore for the name change to azul

from dcc-dashboard-service

jessebrennan committed:
Reformat README.md to wrap at 72 chars

Although this number may seem arbitrary, it's better than before!

jessebrennan committed:
Fix formatting of question prompt

jessebrennan committed:
Reorganize README; add instructions for authorization

This commit also elaborates a bit more on the installation steps.
Also some troubleshooting tips are added. Finally, the AWS credential
prompts are made slightly more clear.

jessebrennan committed:
Generate key for encrypting refresh token

coverbeck commented on boardwalk/conf/boardwalk.config.template:
```diff
@@ -1,6 +1,9 @@
 user_aws_secret_key={{ AWS_SECRET_ACCESS_KEY }}
+project_name={{ PROJECT_NAME }}
+contact_email={{ CONTACT_EMAIL }}
+email_whitelist_name={{ EMAIL_WHITELIST_NAME }}
```
Don't you need to add `refresh_token_encrypt_key` in here as well?

jessebrennan replied:
Whoops! yes.

coverbeck commented on README.md:
```diff
@@ -2,40 +2,94 @@
 
-* [Boardwalk](boardwalk/README.md): our file browsing portal on top of Redwood
+* [Boardwalk](boardwalk/README.md): our file browsing portal on top of
+  Redwood
```
I know this predates your changes (it was something that got overlooked), but can you instead say something like `...on top of the cgp-data-store`. Or maybe `on top of the Azul indexer`, as the UI doesn't directly interact with the data store.

jessebrennan replied:
Done.

coverbeck commented on README.md:
```diff
@@ -2,40 +2,94 @@
-* Ubuntu Server 16.04
-* m5.large
-* 60 GB disk
+For small scale development, a t2.medim with 20GB disks running Ubuntu
```
`medium`

coverbeck commented on README.md:
```diff
@@ -2,40 +2,94 @@
+
+#### Make a routing to the VM
+
+You will need to make a URL to access the VM. On the AWS console, go to
```
This is nitpicky, but you can access the VM via SSH without a URL.... I would rephrase this a little bit, but I'm drawing a blank on how to best say it. :)

jessebrennan replied:
No problem. I get your point. I guess the URL just provides a *consistent* way to access the VM.

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
-* You need to have the name of the [S3 bucket you created earlier](#makebucket) handy as the install script will ask for it. The name you input has to be the exact name you gave it when you created the bucket.
-* All metadata reside in an Elasticsearch database. Make sure you have the domain name of that Elasticsearch instance handy.
-* Have the domain name of the _dos-dss server_ handy.
+Now you can begin installation. You will clone the repository on VM and
```
...clone *this* repository...

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+   You will also be asked for a Google site verfication code. For this
+   field you can just enter `NONE`.
+
+1. The bagit-firecloud-lambda is use for the **Export to Firecloud**
```
`used`

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+1. The [S3 bucket you created earlier](#makebucket) is what you should
+   use for this next step.
+
+1. The dos-dss server is needed next. Again you can use a prexisting one
```
I think you will always use an existing one.

coverbeck replied:
I think you missed this comment.

jessebrennan replied:
I'll link to https://github.com/DataBiosphere/cgp-data-store as we discussed for more context.

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+[here](https://github.com/DataBiosphere/cgp-deployment/blob/feature/update-readme/boardwalk/README.md#development-mode)
+for more details). In addition, the `nginx` image is built from the
+*nginx-dev* directory. If your work requires real SSL certificates
+during development, it is recommended to set up *Common* in `prod` mode,
```
Maybe put this at the very beginning of the section, i.e., even if running in dev mode, we recommend running Common in prod mode. Otherwise it's easy (for me) to overlook, and I think you ran into this too when you set up your first instance.

jessebrennan replied:
Sure, moving makes sense. I wasn't sure what to do because the install script doesn't use the word Common, but instead gateway nginx server. I've since added parentheticals to clarify the redundant names.

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+### <a name="prod"></a> Installing in `prod` mode
 
-### Running the Installer
-Once you have collected the above information, clone the repository on VM and run the bootstrap script. Be sure to set your branch to `feature/commons` as these instruction are specific to this branch.
```
Optional: Maybe there should be a small section on why you would choose dev vs. prod mode. It's sort of in the details, but I don't think it's clear.  Or maybe the names dev and prod make it clear enough?

jessebrennan replied:
Sure, I added something.

coverbeck commented on README.md:
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
 
 * the bootstrapper should install Java, Dockstore CLI
-* Consonance config.template includes hard-coded Consonance token, needs to be generated and written to .env file just like Beni does
+* Consonance config.template includes hard-coded Consonance token, needs
```
Please remove this, it was overlooked. This branch does not include Consonance.

coverbeck commented on install_bootstrap:
```diff
@@ -58,14 +58,14 @@ function generate_password {
-  prev_val=$2
-  val_name=$3
-  answer=''
+    question_str=$1
```
Did you change tabs to spaces or vice-versa, or otherwise do some reformatting?

With exceptions, it's generally best not to add a feature and then also reformat a bunch of the file; it makes it harder to tell what changed.

This comment applies to the whole file, not just this line.

OTOH, if part of your change was just to reformat the indentation, then maybe it's OK.

jessebrennan replied:
I did all of the reformatting in a separate commit https://github.com/DataBiosphere/cgp-deployment/pull/56/commits/20043e98f3e7306450fd98558d2a051671a8d7e5, that way you would be able to tell the changes apart. The indentation was very inconsistent and it was almost necessary to adjust before adding my own code.

coverbeck commented on install_bootstrap:
```diff
@@ -181,9 +181,9 @@ done
-      echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
-      common_mode=-1
-      while [[ "${common_mode}" = "-1" ]]; do
+        echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
```
I'm wondering if we should say recommend prod mode here? What do you think?

jessebrennan replied:
I added a note that explains the limitations for using dev mode here.

coverbeck reviewed (changes requested):
I like your changes to the README, I think it's looking a lot better.

I know some of my feedback goes beyond what you changed, but since you're in there, I think it would be good to address.

jessebrennan committed:
Fix typos and clarify readme

coverbeck commented on README.md:
```diff
@@ -25,7 +25,8 @@ These components are setup with the install process available in this
 * [Boardwalk](boardwalk/README.md): our file browsing portal on top of
-  Redwood
+  the [Azul
+  Indexer](https://github.com/DataBiosphere/azul/tree/feature/commons).
```
👍

coverbeck commented on README.md:
```diff
@@ -80,11 +81,12 @@ instance with 60 GB disk has been sufficient.
-zone you can use, select it from the list. If not, create one to suit
-your needs. Once the zone is selected, click **Create Record Set** at
-the top. Choose a name you like and use the elastic ip as the value.
+You will need to make a URL for consistent access to the VM. On the AWS
```
I'm not sure I like `consistent`. And in hindsight, I don't think I like `make a URL`. How about something like:

You will need to create a DNS record that maps to your VM.

jessebrennan replied:
This is better.

coverbeck commented on install_bootstrap:
```diff
@@ -181,7 +181,8 @@ done
     if [[ $(sudo docker ps | grep 'core-' | wc -l) = "0" ]]; then
-        echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
+        echo "Do you want to run the public-facing gateway nginx server (Common) in dev mode or prod? [dev/prod]"
+        echo "NOTE: dev mode will issue fake certificates, which may be insufficient for developing authentication."
```
👍

coverbeck commented on README.md:
```diff
@@ -44,94 +100,213 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+and *Boardwalk* in `dev` mode.
+
+Once the installer completes, the system should be up and running.
+Congratulations! Execute `docker ps` to get an idea of which containers
```
I think you need to run `sudo docker ps`.

coverbeck reviewed (changes requested):
Sorry, a couple of more tweaks, including one I missed from before.

jessebrennan committed:
More typos and clarification

coverbeck commented on README.md:
```diff
@@ -44,94 +100,213 @@ Open inbound ports on your security group. Use the table below as a guide. Make
+   to access the data.
+
+1. You will now have to enter the hostname of dcc-dashboard. This is the
+   domain you made in Rounte 52, above. You need to know the name of the
```
`Route 53`

coverbeck replied:
And can you make `above` be a link to the section? I was looking for it, but had a hard time finding it because the header says `Make a routing to the VM`. I entered a separate suggestion to rename the title, but even so, it would be nice to have a link as well.

jessebrennan replied:
Sure that would help, I agree.

coverbeck commented on README.md:
```diff
@@ -2,40 +2,96 @@
+    shows you how to ssh into your VM from a terminal.
+
+
+#### Make a routing to the VM
```
Can we call this `Create a DNS entry for the VM`, or `Create a DNS entry for the VM in Route 53`?

jessebrennan replied:
sounds good to me

jessebrennan committed:
Fix typo and add link to route 53 section
